File: ZKCache/src/main/java/com/ericsson/ema/tim/dml/predicate/LessThan.java
package com.ericsson.ema.tim.dml.predicate;

import com.ericsson.ema.tim.dml.DataTypes;
import com.ericsson.ema.tim.exception.DmlBadSyntaxException;
import com.ericsson.ema.tim.exception.DmlNoSuchFieldException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;

/**
 * Created by eqinson on 2017/4/23.
 */
public class LessThan extends AbstractPredicate implements Predicate {
    private final static Logger LOGGER = LoggerFactory.getLogger(LessThan.class);

    private LessThan(String field, int value) {
        super(field, value);
    }

    public static LessThan lt(String field, int value) {
        return new LessThan(field, value);
    }

    @Override
    public boolean eval(Object tuple) {
        if (this.valueToComp == null)
            return false;

        Object fieldVal = getFiledValFromTupleByName(tuple);
        Map<String, String> metadata = getSelector().getContext().getTableMetadata();
        String fieldType = metadata.get(field);
        if (fieldType == null)
            throw new DmlNoSuchFieldException(field);

        switch (fieldType) {
            case DataTypes.Int:
                return Integer.compare((Integer) fieldVal, (Integer) this.valueToComp) < 0;
            default:
                LOGGER.error("must be int type in LessThan: {},{}", field, fieldType);
                throw new DmlBadSyntaxException("must be int type in LessThan: " + field + "," + fieldType);
        }
    }
}

File: zkcache_scala/src/main/scala/com/ericsson/ema/tim/dml/predicate/LessThan.scala
package com.ericsson.ema.tim.dml.predicate

import com.ericsson.ema.tim.dml.DataTypes
import com.ericsson.ema.tim.exception.{DmlBadSyntaxException, DmlNoSuchFieldException}
import org.slf4j.LoggerFactory

/**
  * Created by eqinson on 2017/5/13.
  */
class LessThan private(override val field: String, override val valueToComp: Object) extends AbstractPredicate(field,
	valueToComp) with Predicate {
	private val LOGGER = LoggerFactory.getLogger(classOf[LessThan])

	override def eval(tuple: Object): Boolean = {
		if (Option(this.valueToComp).isEmpty)
			return false

		val fieldVal: Object = getFiledValFromTupleByName(tuple)
		val fieldType = selector.context.tableMetadata.get(field)
		fieldType match {
			case Some(DataTypes.Int) =>
				fieldVal.asInstanceOf[Integer].compareTo(valueToComp.asInstanceOf[Integer]) < 0
			case Some(other)         =>
				LOGGER.error("unsupported data type: {},{}", field, other: Any)
				throw DmlBadSyntaxException("unsupported data type: " + field + "," + other)
			case None                => throw DmlNoSuchFieldException(field)
		}
	}
}

object LessThan {
	def apply(field: String, value: Int) = new LessThan(field, Integer.valueOf(value))
}
